### PR TITLE
dbeaver/dbeaver#20592 keyboard shortcut set row color

### LINKED ISSUE #20592: Keyboard shortcut for "Set row color for..." in results
First, I **love** dbeaver!  And just how hard all of you work to make it even better.  Thank you!!  :-)

**Is your feature request related to a problem? Please describe.**
Easily setting the row color for a value is VERY handy when looking at lots of code.  I tend to use this a lot.  But every time it is:
- right-click
- view/format
- Set row color for...

**Describe the solution you'd like**
Please add an option in the keyboard shortcuts to go straight to "Set row color for...".  This would automatically bring up the color selector, you select a color and done.

Even AWESOMER :-) would be having multiple keyboard shortcuts for setting the row color to different colors.  But I understand where that could/would get complicated quickly.  <wry smile>

"Solution" 2:
Adding an option in the keyboard shortcuts would work best for everyone, but **I** could get by with a letter "default" (the letter that is underlined in a menu that you can just hit to go to that menu item, sorry don't remember the "proper" name for them) for the "Set row color for..." entry with my Autohotkey script (see below).   

"the letter that is underlined in a menu..." - example, in the Window menu, it shows Editor with the "E" being underlined.

**Describe alternatives you've considered**
I use Autohotkey for lots of things and _tried_ to create a hotkey to effectively do the right-click, v (for view/format), then hitting "s" a couple of times to get down to "Set row color for..." since there are no letter "defaults" (or whatever they are called).  

HOWEVER, some columns show all three below that respond to "s", and others don't have the first one (Set ... transform), so it doesn't really work.  <wry smile>
- Set "<column name>" Transform
- Show complex column structures
- Set row color for...

And hitting "r" instead of "s" takes me right into "Customize colors for <query>", which is much more detailed, but also a lot more to do for just setting a specific color for a value in this result set.

**Additional context**
image

### PR DESCRIPTION
**Pull request by serjiokov**
issue: Keyboard shortcut for "Set row color for..." in results #20592

Provided specific handler in order to set up the row color binding by user choise.

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java [hunk 1]
```diff
@@ -21,41 +21,110 @@
...
-import org.eclipse.jface.action.*;
+import org.eclipse.jface.action.Action;
+import org.eclipse.jface.action.ActionContributionItem;
+import org.eclipse.jface.action.GroupMarker;
+import org.eclipse.jface.action.IAction;
+import org.eclipse.jface.action.IContributionItem;
+import org.eclipse.jface.action.IMenuManager;
+import org.eclipse.jface.action.MenuManager;
+import org.eclipse.jface.action.Separator;
+import org.eclipse.jface.action.ToolBarManager;
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java]: Please return this back as it was. I believe these changes are unintended.
We use import with * when there is more than 4 classes of the same package.

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java [hunk 2]
```diff
@@ -2944,7 +3061,15 @@ public void run() {
...
             if (valueController != null) {
-                viewMenu.add(new SetRowColorAction(this, attr, valueController.getValue()));
+                String bindingText = attr.getName() 
+                        + " = "
+                        + UITextUtils.getShortText(this.getControl(),
+                        CommonUtils.toString(valueController.getValue()), 100);
+                String msgSelectRowColor = NLS.bind(ResultSetMessages.actions_name_color_by, bindingText);
+                // select row color
+                viewMenu.add(ActionUtils.makeCommandContribution(this.getSite(),
+                        "org.jkiss.dbeaver.core.resultset.grid.selectRowColor",
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java]: Please use a constant you added CMD_SELECT_ROW_COLOR

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java [hunk 3]
```diff
@@ -2944,7 +3061,15 @@ public void run() {
         viewMenu.add(new Separator());
         if (model.getDocumentAttribute() == null) {
             if (valueController != null) {
-                viewMenu.add(new SetRowColorAction(this, attr, valueController.getValue()));
+                String bindingText = attr.getName() 
+                        + " = "
+                        + UITextUtils.getShortText(this.getControl(),
+                        CommonUtils.toString(valueController.getValue()), 100);
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java]: We use 4-spaces indentation always.
```suggestion
                String bindingText = attr.getName() 
                    + " = "
                    + UITextUtils.getShortText(this.getControl(),
                    CommonUtils.toString(valueController.getValue()), 100);
```

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java [hunk 4]
```diff
@@ -553,17 +564,46 @@ protected IStatus run(DBRProgressMonitor monitor) {
...
+                            final DBDAttributeBinding attr = rsv.getActivePresentation().getCurrentAttribute();
+                            ResultSetCellLocation currentCellLocation = ssp.getCurrentCellLocation();
+                            Object cellValue = resultSetViewer.getContainer().getResultSetController().getModel()
+                                    .getCellValue(currentCellLocation);
+                            vEntity.setColorOverride(attr, cellValue, null, StringConverter.asString(color));
+                            updateColors(resultSetViewer, vEntity, true);
+                        } catch (IllegalStateException e) {
+                            DBWorkbench.getPlatformUI().showError(
+                                    ResultSetMessages.dialog_row_colors_error_message_title,
+                                    ResultSetMessages.dialog_row_colors_error_message_text, e);
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java]:
```suggestion
                            DBWorkbench.getPlatformUI().showError(
                                ResultSetMessages.dialog_row_colors_error_message_title,
                                ResultSetMessages.dialog_row_colors_error_message_text, e);
```

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java [hunk 5]
```diff
@@ -553,17 +564,46 @@ protected IStatus run(DBRProgressMonitor monitor) {
...
+                            }
+                        } finally {
+                            UIUtils.disposeCenteredShell(shell);
+                        }
+                        try {
+                            final DBVEntity vEntity = getColorsVirtualEntity(resultSetViewer);
+                            final DBDAttributeBinding attr = rsv.getActivePresentation().getCurrentAttribute();
+                            ResultSetCellLocation currentCellLocation = ssp.getCurrentCellLocation();
+                            Object cellValue = resultSetViewer.getContainer().getResultSetController().getModel()
+                                    .getCellValue(currentCellLocation);
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java]:
```suggestion
                            Object cellValue = resultSetViewer.getContainer().getResultSetController().getModel()
                                .getCellValue(currentCellLocation);
```

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java (hunk 2 above)
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java]:
```suggestion
                    "org.jkiss.dbeaver.core.resultset.grid.selectRowColor",
```

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java (hunk 1 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java]: Generaly, it is not a good idea to export all classes from the package.

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java (hunk 2 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java]: Nice catch

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java (hunk 1 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java]: There is a nice link describing pros and cons of managing includes https://www.baeldung.com/java-wildcard-imports.
I'm voating to define includes obviusly and avoid any conflicting dependency of wildcard using, at same time standard operation of Ctrl+O will work perfectly

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java (hunk 4 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java]: applied

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java (hunk 2 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java]: applied

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java (hunk 5 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java]: applied

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java (hunk 1 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java]: I'm not deciding for the whole team, just telling you our current rules.

Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetViewer.java]: Ok, get it, will return.

File: plugins/org.jkiss.dbeaver.ui.editors.data/plugin.xml [hunk 6]
```diff
@@ -1015,7 +1018,11 @@
         <key commandId="org.jkiss.dbeaver.core.resultset.grid.hideColumns" contextId="org.jkiss.dbeaver.ui.context.resultset" schemeId="org.jkiss.dbeaver.keyboardOnlyKeyScheme" sequence="ALT+SHIFT+H"/>
         <key commandId="org.jkiss.dbeaver.core.resultset.grid.showColumns" contextId="org.jkiss.dbeaver.ui.context.resultset" schemeId="org.jkiss.dbeaver.keyboardOnlyKeyScheme" sequence="ALT+SHIFT+T"/>
         <key commandId="org.jkiss.dbeaver.core.resultset.grid.showColumnContextMenu" contextId="org.jkiss.dbeaver.ui.context.resultset" schemeId="org.jkiss.dbeaver.keyboardOnlyKeyScheme" sequence="SHIFT+F11"/>
-
+        <key
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ui.editors.data/plugin.xml]: We do not need additional shortcut. Lets users assign it by themselves.

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.ui.editors.data/OSGI-INF/l10n/bundle.properties [hunk 7]
```diff
@@ -71,6 +71,11 @@ command.org.jkiss.dbeaver.core.resultset.grid.selectColumn.name=Select column(s)
 command.org.jkiss.dbeaver.core.resultset.grid.selectColumn.description=Select columns for the selected cell(s)
 command.org.jkiss.dbeaver.core.resultset.grid.selectRow.name=Select row(s)
 command.org.jkiss.dbeaver.core.resultset.grid.selectRow.description=Select row(s) for the selected cell(s)
+
+command.org.jkiss.dbeaver.core.resultset.grid.selectRowColor.name=Set row color
+command.org.jkiss.dbeaver.core.resultset.grid.selectRowColor.description=Set color for the selected row
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ui.editors.data/OSGI-INF/l10n/bundle.properties:75]: Please remove redundant empty lines

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.ui.editors.data/plugin.xml (hunk 6 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui.editors.data/plugin.xml]: Done

File: plugins/org.jkiss.dbeaver.ui.editors.data/OSGI-INF/l10n/bundle.properties (hunk 7 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui.editors.data/OSGI-INF/l10n/bundle.properties:75]: Done

**Review by E1izabeth — APPROVED** [review]

**Comment by Matvey16** [comment]
There is no binding by default

image

**Comment by E1izabeth** [comment]
There should be no default binding.
@serjiokov please update PR description :)

**Comment by Matvey16** [comment]
Verified

**Review by serge-rider — APPROVED** [review]

**Pull request merged**